Plan bind

lakshmi-somasundaram opened the pull request:
Adding plan bind to Linkedit groovy 
Changing the binds to JCLEXEC to accomodate remote DB2 locations on sysplex

dajsen commented:
@lakshmi-somasundaram 
I think I understand the bind/bindPackage/bindPlan definitions in plan bind changes.
Don't understand the use of bindP definition though, When is that one executed ?
Thanks.

drbruce-git commented:
@lakshmi-somasundaram can you follow the instructions listed in [CONTRIBUTIONS.md](https://github.com/IBM/dbb/blob/master/CONTRIBUTIONS.md) so that we can consider your pull request?

lakshmi-somasundaram commented:
bindP definition is called from Linkedit.groovy when bind_performBindPlan is true. I might have missed the bind.properties in this let me know, I have do a new pull request

dajsen commented:
@lakshmi-somasundaram 
Hi.
regarding bindP, I think you are wrong (will talk to some groovy-people today to figure it out).
regarding bind.properties, missing file in your pull request since you have added some new stuff?
regarding the call in LinkEdit.groovy: 
bindUtils.bindPlan(buildFile, props.buildOutDir... Where is bindUtils defined ? I think that you have to put something like this at the beginning of LinkEdit.groovy: 
@Field def bindUtils= loadScript(new File("${props.zAppBuildDir}/utilities/BindUtilities.groovy"))
Compare with this one, which is defined at startup: 	
@Field def buildUtils= loadScript(new File("${props.zAppBuildDir}/utilities/BuildUtilities.groovy"))
// copy build file to input data set
buildUtils.copySourceFiles(buildFile, props.linkedit_srcPDS, null, null)

lakshmi-somasundaram commented:
regarding bindP, I think you are wrong. This one will never get executed unless if you make a call to that particular method.
regarding bind.properties, missing file in your pull request since you have added some new stuff?
regarding the call in LinkEdit.groovy:
bindUtils.bindPlan(buildFile, props.buildOutDir... Where is bindUtils defined ? I think that you have to put something like this at the beginning of LinkEdit.groovy:
@field def bindUtils= loadScript(new File("${props.zAppBuildDir}/utilities/BindUtilities.groovy"))
Compare with this one, which is defined at startup:
@field def buildUtils= loadScript(new File("${props.zAppBuildDir}/utilities/BuildUtilities.groovy"))
>>>> I have selectively copied thats why it was missing the load script step, its working on our side 
bindP is called on Linkedit.groovy

Below are the additional properties added for Plan bind 
# Flag indicating to run DB2 PLAN BIND.  Set to 'true' to bind SQL programs
bind_performBindPlan=false

# Specify the owner of the plan
bind_planOwner=

dajsen commented:
@lakshmi-somasundaram 
Looks Great!
bindP/bindPlan not the same thing...  Don't agree on that one, but it's ok. 
property bind_planOwner not needed or using it somewhere else ? You already have 'owner'.
def (bindRc, bindLogFile) = bindUtils.bindPlan(buildFile, props.buildOutDir, props.bind_runIspfConfDir, props.bind_db2Location, props.bind_collectionID, owner, props.bind_qualifier, props.verbose && props.verbose.toBoolean());

lakshmi-somasundaram commented:
> @lakshmi-somasundaram
> Looks Great!
> bindP/bindPlan not the same thing... Don't agree on that one, but it's ok.
> property bind_planOwner not needed or using somewhere else ? You already have 'owner'.
> def (bindRc, bindLogFile) = bindUtils.bindPlan(buildFile, props.buildOutDir, props.bind_runIspfConfDir, props.bind_db2Location, props.bind_collectionID, owner, props.bind_qualifier, props.verbose && props.verbose.toBoolean());

>>might be a miss on my part on the owner, bind package owner and bind plan owners can have different values
>>I didnt get the bindP/bindPlan - but just a note, JCL inside plan and package are different they are not the same

dajsen commented:
OK, you are only using one of the values. I know that bindPlan/bindPackage are different. 
How do you solve 'grant execute on plan to public' in your shop?
Review coming up...

lakshmi-somasundaram commented:
> OK, you are only using one of the values. I know that bindPlan/bindPackage are different.
> How do you solve 'grant execute on plan to public' at your shop?
> Review coming up...

We dont grant to public, grants are issued/executed as JCL separately, it can be added as a next step to JCL if needed

dajsen commented:
OK, similar problem maybe. We don't have the proper priviliges to do that in dbb yet, so executed outside of dbb due to security issues...

dajsen commented:
Proc DSNTSOP is most likely local (Statefarm). Maybe have to change to a more common proc before merging ?

lakshmi-somasundaram commented:
> Proc DSNTSOP is most likely local (Statefarm). Maybe change to a more common proc before merging ?

Will DSNTEPE work, let me check with our DB2 support on it

lakshmi-somasundaram commented:
> Hi,
> There are some changes that need to be done before we can consider merging into master.
> 
> **LinkEdit.groovy**
> 
> * Can you set `bindFlag ` to `false` if the Link Edit step failed.
> 
> ```
> 	if (rc > maxRC) {
> 	    bindFlag = false
> ```
> 
> Do you think that having only `bind_maxRC` for the two kind of bind is enough. We may need a separate `bindPlan_maxRc` ?
> 
> I was not able to test at runtime. The generated JCL failed for me and this case is not handled by your code.
> 
> When a JCL is in error, this API `exec.maxRC` return `JOB ERROR`. So the calling this code ` exec.maxRC.split("CC")[1].toInteger()` will generate an exception.
> 
> ```
> java.lang.ArrayIndexOutOfBoundsException: Array index out of range: 1
>         at org.codehaus.groovy.runtime.dgmimpl.arrays.ObjectArrayGetAtMetaMethod.invoke(ObjectArrayGetAtMetaMethod.java:41)
>         at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
>         at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.invoke(PojoMetaMethodSite.java:51)
>         at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:56)
>         at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
>         at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
>         at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
> ```
> 
> If I quickly code this to go deeper:
> 
> ```
>     def jobRc= exec.maxRC.split("CC")
> 	int rc = Integer.MAX_VALUE
> 	if ( jobRc.length > 1 )
> 		rc = exec.maxRC.split("CC")[1].toInteger() 	
> 	println "***Bind Job ${exec.submittedJobId} completed with $rc "
> 	exec.saveOutput(logFile)
> ```
> 
> Here the reason why the JCL is in error:
> 
> ```
>   STMT NO. MESSAGE
>          2 IEFC612I PROCEDURE DSNTSOP WAS NOT FOUND
>          2 IEFC657I THE SYMBOL RC WAS NOT USED
> ```
> 
> So can you review this ? Is there are some values that must be specific to the system these values must go in the property file. Example here `DSNTSOP`.
> 
> Regards

I shall make the update and send it across

dajsen commented:
@lakshmi-somasundaram 
You know what happens when you change stuff in a hurry, not testing and so on... 
Better safe than sorry, take your time! 
Have a look at the bind.props: bind_planMaxRC not set...
What about bind_performBindPackage=false, bind_performBindPlan=false. When do you set those to true ?

jbyibm commented:
> What about bind_performBindPackage=false, bind_performBindPlan=false. When do you set those to true ?

For me it is something we set to `true` in the property file manually. Depending on customer configuration we may want to not bind anything.

dajsen commented:
@jbyibm 
'For me it is something we set to true in the property file manually. ' 
They are set to true in the props-file if the customer runs Db2 ? If not, then I have the wrong setup...
Since it's an application value, it depends on the application...

jbyibm commented:
The binding in the current implementation will run only during an IDE user build (flag `props.userBuild`). User build is generally triggered by the developer for many reasons:
- Just to test if the modifications he made can compile. In this case he can not be interested in the binding step.
- He can want to test its modification on a development environment. So he will have to do the bind.

In all cases it is related to Db2

dajsen commented:
@jbyibm 
in bind.utilities you have this method: bindP (which is sort of a copy of the bind method).
When and how is that one executed and what do you need it for ?

jbyibm commented:
> @jbyibm
> in bind.utilities you have this method: bindP (which is sort of a copy of the bind method).
> When and how is that one executed and what do you need it for ?

I think `bind` and `bindP` should be merged into one method to allows to call the groovy file directly (main method). This new method should have arguments for both bindPlan/bindPackage.

But before that we must be sure to have something that works on all system. This is actually the case in the original implementation. We need a replacement for `DSNTSOP`.

I'm just reviewing the proposal code here.

lakshmi-somasundaram commented:
I was halfway through the changes, was trying to test on my local and got held up in between, my changes are not complete yet, also bind and bindP has different arguments,  dbrmlib is not needed for bindP

lakshmi-somasundaram commented:
Added my changes, I have left the max rc as 16,  if 16 its a bad return code anyway.  Pulled in the generic job for bind, still might need some load lib changes based on each shop. bind and bindP has different input parms, bindP does need DBRMLIB as input thats why its separated as 2 different functions.

jbyibm commented:
@lakshmi-somasundaram Thanks for the update. I will look later today.

lakshmi-somasundaram commented on Build/zAppBuild/utilities/BindUtilities.groovy:
```diff
@@ -12,69 +12,161 @@ bind(args)
+       OWNER(${OWNER})       +                                
+       QUALIFIER(${QUAL})    +                                
+       ACTION(REPLACE)     +   
+       LIBRARY('${dbrmHLQ}')
```
Added the updates

jbyibm commented:
@lakshmi-somasundaram I will review last commit ASAP.

drbruce-git commented:
@lakshmi-somasundaram @jbyibm Thanks for working through the issues on this contribution.  A couple of points:
- We have created a stand alone repository for zAppBuild at https://github.com/IBM/dbb-zappbuild.  This is to allow for easier delivery of maintenance and enhancements to users and customers who are using it as part of their zDevOps pipeline solutions.  As such we will be removing zAppBuild from this repository at the end of this week on March 13 to align with the release of DBB v1.0.8 when we normally update the DBB samples.  This being the case we will no longer be accepting pull requests for zAppBuild against this repository. Please recreate this pull request against the new repository
- Since we are now hosting zAppBuild in its own repository specifically for delivering maintenance and enhancements to customers using zAppBuild as part of their zDevOps solution, we need to be careful about backward compatibility and breaking existing customer applications. While changes in the build scripts that reference changes in build-conf are probably safe since they would be self contained in zAppBuild versions, changes to application-conf files are potentially dangerous. We need to be able to test that any new function added that requires changes to application-conf won't cause problems if those properties are absent.  @jbyibm noted that this pull request at one point would break backward compatibility.  Are any of these changes in application-conf?  If so we need to figure out a way to maintain backward compatibility.

jbyibm commented:
Thanks @drbruce-git for your comments. 
I will to my best to review the latest commit code this week and propose a solution for backward compatibility. Then I will send my code to @ lakshmi-somasundaram so that it can create a PR under a single and full commit against the new repository.

jbyibm commented:
I was very busy this week.  I just start the review now.

jbyibm commented:
I have whole code. When I will have the reviewed the code, I will ping @lakshmi-somasundaram.  @lakshmi-somasundaram  will be able to create a new PR on the new repository to keep ownership of the code.

drbruce-git commented:
Closing this pull request now that zAppBuild has been moved to https://github.com/IBM/dbb-zappbuild

jbyibm commented:
@lakshmi-somasundaram Sorry for the delay but I had difficulty to find time to work on this.

Here the steps to take so you can commit the new code in the new repository:

1. Fork the new repo https://github.com/IBM/dbb-zappbuild
2. On your fork in the `development` branch replace the files included in the archive here.
3. Performs some tests if possible in your environment
4. Create a new issue in this repo https://github.com/IBM/dbb-zappbuild
5. Create a PR against the `development`. Do your best to only have 1 commit as recommended here https://github.com/IBM/dbb-zappbuild/blob/development/CONTRIBUTIONS.md

[patch.tar.gz](https://github.com/IBM/dbb/files/4360124/patch.tar.gz)

jbyibm commented:
@lakshmi-somasundaram Do you any chance to look at my last comment ?

lakshmi-somasundaram commented:
I have forked the new one, need to check on the process for downloading the tar file from our network

jbyibm commented:
@lakshmi-somasundaram can you take into account this PR when you will merge https://github.com/IBM/dbb-zappbuild/pull/14? 
Thanks.
